cmd/tsidp/tsidp.go

```diff
@@ -611,6 +758,124 @@ func (s *idpServer) serveOpenIDConfig(w http.ResponseWriter, r *http.Request) {
 	}
 }
 
+// funnelClient represents an OIDC client/relying party that is accessing the
+// IDP over Funnel.
+type funnelClient struct {
+	ID     string `json:"client_id"`
+	Secret string `json:"client_secret,omitempty"`
+	Name   string `json:"name,omitempty"`
+}
+
+// /clients is a privileged endpoint that allows the visitor to create new
+// Funnel-capable OIDC clients, so it is only accessible over the tailnet.
+func (s *idpServer) serveClients(w http.ResponseWriter, r *http.Request) {
+	if isFunnelRequest(r) {
+		http.Error(w, "tsidp: not found", http.StatusNotFound)
+		return
+	}
+
+	path := strings.TrimPrefix(r.URL.Path, "/clients/")
+
+	if path == "new" {
+		s.serveNewClient(w, r)
+		return
+	}
+
+	if path == "" {
+		s.serveGetClientsList(w, r)
+		return
+	}
+
+	s.mu.Lock()
+	c, ok := s.funnelClients[path]
+	s.mu.Unlock()
+	if ok {
+		switch r.Method {
+		case "DELETE":
+			s.serveDeleteClient(w, r, path)
+		case "GET":
+			json.NewEncoder(w).Encode(c)
+		default:
+			http.Error(w, "tsidp: method not allowed", http.StatusMethodNotAllowed)
+		}
+		return
+	}
+
+	http.Error(w, "tsidp: not found", http.StatusNotFound)
```

Comment:
prefer to indent the error flow

https://go.dev/wiki/CodeReviewComments#indent-error-flow